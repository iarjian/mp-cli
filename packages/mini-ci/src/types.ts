import { IInnerUploadOptions } from 'miniprogram-ci/dist/@types/ci/upload';

export interface ReplaceRuleItem {
	search: string | Function;
	replace: string | Function;
}

export interface ReplaceRules {
	file: string;
	rules: ReplaceRuleItem[];
}

export interface Plugin {
	enforce?: 'pre' | 'post';
	handler: Function;
	name: string;
}

export interface UserConfig {
	robot?: number;
	root?: string;
	name?: string;
	projectPath: string;
	packageJsonPath?: string;
	appid: string;
	privateKeyPath: string;
	// # 多个小程序配置的路径
	configPath?: string;
	// # 版本描述可选，默认会读取版本号和git commit hash 和作者
	// # 如：v1.0.0 - 4c29ae8 - by@若川
	desc?: string;
	// # 版本号可选，默认读取 package.json 中的 version
	version?: string;
	replaceRules?: ReplaceRules[];
	// 参考文档：https://developers.weixin.qq.com/miniprogram/dev/devtools/ci.html
	// 其他项目配置自定义配置，和上面配置合并
	projectOptions?: {};
	// 新增上传的自定义配置
	uploadOptions?: IInnerUploadOptions;
	// 新增下载的自定义配置
	previewOptions?: {};
}

export interface InlineConfig {
	// @default process.cwd()
	root?: string;
	useMultiSelect?: boolean;
	useSelect?: boolean;
	robot?: number;
	dry?: boolean;
	useAllConfig?: boolean;
}
