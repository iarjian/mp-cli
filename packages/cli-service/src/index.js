const Koa = require('koa');
const Router = require('@koa/router');
const path = require('path');
const cors = require('@koa/cors');
const bodyParser = require('koa-bodyparser');
const { execSync } = require('child_process');
const fs = require('fs');

const app = new Koa();
const router = new Router();

app.use(cors());
app.use(bodyParser());

router.get('/', (ctx, next) => {
	// ctx.router available
	ctx.body = '这里是主页';
});

router.post('/upload', (ctx, next) => {
	const {
		body: { basename },
	} = ctx.request;
	// ctx.router available
	console.log('project', ctx.request, basename);
	const result2 = execSync(
		'node ./publish/index.js --notSelect --config ' + basename,
	);
	ctx.body = {
		code: 1,
		// data: body,
		data: result2.toString(),
		msg: '上传成功',
	};
	// await next();
});

router.get('/list', (ctx, next) => {
	// ctx.router available
	const path2 = path.join(process.cwd(), '../../src/config');
	const configDir = fs.readdirSync(path2);
	const config = configDir.map((el) => {
		const json = require(path2 + '/' + el);
		return {
			...json,
			basename: el,
			appid: json.appid,
			url: json.url || '-',
			name: json.name || '-',
		};
	});
	ctx.body = {
		code: 1,
		data: config,
		msg: '获取成功',
	};
});

app.use(router.routes()).use(router.allowedMethods());

app.listen(3002, () => {
	console.log('端口监听在3002');
});
