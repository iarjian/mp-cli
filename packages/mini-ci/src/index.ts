export { default as logger } from './logger';

export * from './types';

export { resolveConfig } from './config';

export { generatorActions, preview, upload, init } from './commands';
