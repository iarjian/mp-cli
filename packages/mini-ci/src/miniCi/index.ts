import WeappCI from './weappCi';

export const miniCi = (options, platform: string = 'weapp') => {
	const platformMap = {
		weapp: WeappCI,
	};
	let PlatformCi = platformMap[platform];
	if (!PlatformCi) {
		throw new Error(`暂时不支持 ${platform} 平台`);
	}
	const ci = new PlatformCi(options);
	return ci;
};
