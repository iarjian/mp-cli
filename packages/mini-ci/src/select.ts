import fs from 'node:fs';
import enquirer from 'enquirer';
import logger from './logger';
import { InlineConfig, UserConfig } from './types';

interface SelectOptions {
	configList: any[];
	useSelect?: InlineConfig['useSelect'];
	useMultiSelect?: InlineConfig['useMultiSelect'];
}

export const select = async (options: SelectOptions) => {
	let {
		configList,
		useSelect = false,
		useMultiSelect = false,
	} = options || {};

	// logger.log('根据单选或者多选选择小程序参数', options);

	// 单选、多选 两者只能选择其一
	if (useSelect) {
		useMultiSelect = false;
	}
	if (useMultiSelect) {
		useSelect = false;
	}

	const { prompt, MultiSelect } = enquirer;

	// logger.log('configList', configList);

	let result: UserConfig[] = [];

	if (useSelect) {
		const { name } = await prompt({
			type: 'select',
			name: 'name',
			message: '请选择一个小程序配置',
			choices: configList,
		});
		result = configList.filter((el) => el.name === name);
		return result;
	}

	if (useMultiSelect) {
		const multiSelectPrompt = new MultiSelect({
			name: 'value',
			message: '可选择多个小程序配置',
			limit: 7,
			choices: configList,
		});

		try {
			const answer = await multiSelectPrompt.run();
			logger.log('您已选择:', answer);
			result = configList.filter((el) => answer.includes(el.name));
			return result;
		} catch (err) {
			logger.log('您已经取消', err);
			process.exit(1);
		}
	}

	return result;

	// const { yes } = await prompt({
	// 	type: 'confirm',
	// 	name: 'yes',
	// 	message: `确认选择这(几)个小程序？`,
	// });

	// if (!yes) {
	// 	return;
	// }
};
