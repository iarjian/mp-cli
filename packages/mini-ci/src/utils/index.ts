import os from 'node:os';
import path from 'node:path';
import fs from 'node:fs';
import { InlineConfig } from '../types';
export { default as copy } from './copy';

export const isObject = (val: any) => typeof val === 'object' && val !== null;

export const readJSON = (file: fs.PathOrFileDescriptor) =>
	JSON.parse(fs.readFileSync(file, 'utf8'));

export function slash(p: string): string {
	return p.replace(/\\/g, '/');
}

// const pkg = readJSON(new URL('../package.json', import.meta.url));

export const isWindows = os.platform() === 'win32';
export function normalizePath(id: string): string {
	return path.posix.normalize(isWindows ? slash(id) : id);
}

export const getResolvedRoot = (config: InlineConfig) => {
	return normalizePath(
		config.root ? path.resolve(config.root) : process.cwd(),
	);
};

// 简单的转数组的函数
export const arrify = (arr: any) => {
	if (Array.isArray(arr)) {
		return arr;
	} else {
		return [];
	}
};

const camelizeRE = /_(\w)/g;
/**
 * @private
 */
export const camelize = (str: string): string => {
	return str.replace(camelizeRE, (_, c) => (c ? c.toUpperCase() : ''));
};

export const isString = (val: any) => typeof val === 'string';
