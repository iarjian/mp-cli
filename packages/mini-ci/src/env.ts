import path from 'node:path';
import dotenv, { DotenvParseOutput } from 'dotenv';
import dotenvExpand from 'dotenv-expand';
import { camelize, isString } from './utils';

// 加载 .env 配置
export const loadEnv = (
	mode: string | undefined,
	envDir: string,
	prefix: string = 'MINI_',
) => {
	let env: Record<string, string> = {};
	let envFile = '.env';

	if (isString(mode)) {
		envFile = `.env.${mode}`;
	}

	let dotenvResult = dotenv.config({
		path: path.resolve(envDir, envFile),
	});

	if (dotenvResult.error) {
		throw dotenvResult.error;
	}

	const dotenvExpandResult = dotenvExpand.expand({
		parsed: dotenvResult.parsed,
		ignoreProcessEnv: true,
	});

	if (dotenvExpandResult.error) {
		throw dotenvExpandResult.error;
	}

	// @ts-ignore
	for (const [key, value] of Object.entries(dotenvExpandResult.parsed)) {
		if (key.startsWith(prefix)) {
			const lastKey = camelize(key.replace(prefix, '').toLowerCase());
			env[lastKey] = value;
		}
	}

	return env;
};
