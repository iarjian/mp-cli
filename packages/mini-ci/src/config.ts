import path from 'node:path';
import fs from 'node:fs';
import { execaCommandSync } from 'execa';
import { pathToFileURL } from 'node:url';

import { getResolvedRoot, isObject, readJSON, uniqueId } from './utils/index';
import { DEFAULT_CONFIG_FILES } from './const';
import { loadEnv } from './env';
import logger from './logger';
import { select } from './select';
import { InlineConfig, UserConfig } from './types';

// 获取 package.json 版本号
export const getVersion = (packageJsonPath: string) => {
	let version = '1.0.0';
	try {
		version = readJSON(`${packageJsonPath}`).version;
	} catch (e) {
		logger.warn(
			'未设置 version , 并且未设置 package.json 路径，无法读取 version，默认使用 1.0.0',
			e,
		);
	}
	return version;
};

// 获取配置
export const getDesc = (
	projectPath: UserConfig['projectPath'],
	version: UserConfig['version'],
) => {
	// 获取最新 git 记录 7位的 commit hash
	let gitCommitHash = 'git commit hash 为空';
	try {
		gitCommitHash = execaCommandSync('git rev-parse --short HEAD', {
			stdio: 'pipe',
			cwd: projectPath,
		}).stdout;
	} catch (e) {
		logger.warn('获取 git commit hash 失败', e);
	}

	// 获取项目的git仓库的 user.name
	let userName = '默认';
	try {
		userName = execaCommandSync('git config user.name', {
			stdio: 'pipe',
			cwd: projectPath,
		}).stdout;
	} catch (e) {
		logger.warn('git config user.name 获取失败', e);
	}

	const desc = `v${version} - ${gitCommitHash} - by@${userName}`;
	return desc;
};

// 获取默认的路径
const getDefaultConfigFilePath = (configDir: string) => {
	// 判断路径
	let resolvedPath: string = '';
	for (const filename of DEFAULT_CONFIG_FILES) {
		const filePath = path.resolve(configDir, filename);
		if (!fs.existsSync(filePath)) continue;

		resolvedPath = filePath;
		break;
	}

	return resolvedPath;
};

// 解析 js/json 配置文件
export const resolveFileConfig = async (filePath: string) => {
	let fileNameTmp: string = '';
	// json 文件
	if (/\.json$/.test(filePath)) {
		return readJSON(filePath);
	}
	try {
		let fileUrl = filePath;
		// 统一用 esm 处理
		// 不是 mjs 结尾，生成临时的 mjs 来读取，读取完后删除
		if (!/\.mjs$/.test(filePath)) {
			const fileBase = `${filePath}.timestamp-${Date.now()}`;
			fileNameTmp = `${fileBase}.mjs`;
			fileUrl = `${pathToFileURL(fileBase)}.mjs`;
			const code = fs.readFileSync(filePath, 'utf8');
			fs.writeFileSync(fileNameTmp, code, 'utf8');
		}

		const res = await import(fileUrl);
		return res.default;
	} catch (e) {
		logger.log(`加载配置文件 ${filePath} 失败`, e);
		return;
	} finally {
		try {
			fs.unlinkSync(fileNameTmp);
		} catch {
			// already removed if this function is called twice simultaneously
		}
	}
};

// 加载配置
const loadConfigFromFile = async (
	configDir: string,
): Promise<UserConfig | undefined> => {
	const resolvedPath = getDefaultConfigFilePath(configDir);
	if (!resolvedPath) {
		return;
	}
	return resolveFileConfig(resolvedPath);
};

// 合并配置
export const mergeConfig = (
	defaults: Record<string, any>,
	config: Record<string, any>,
) => {
	// 获取配置
	const getLastOptions = (val: any) => (isObject(val) ? val : {});
	const projectOptions = {
		projectPath: config.projectPath,
		type: 'miniProgram',
		appid: config.appid,
		privateKeyPath: config.privateKeyPath,
		ignores: [
			'node_modules/**/*',
			'CHANGELOG.md',
			'README.md',
			'yarn.lock',
			'package-lock.json',
			'pnpm-lock.yaml',
		],
		...getLastOptions(config.projectOptions),
	};

	let setting = {};
	try {
		const projectConfig = readJSON(
			`${config.projectPath}/project.config.json`,
		);
		setting = projectConfig?.setting;
	} catch (e) {
		logger.warn(
			`加载项目中的 ${config.projectPath}/project.config.json 失败`,
		);
	}
	const commonConfig = {
		version: config.version || getVersion(config.packageJsonPath),
		robot: defaults.robot,
		setting,
		// onProgressUpdate: console.log,
	};

	const uploadOptions = {
		...commonConfig,
		desc: config.desc || getDesc(config.projectPath, commonConfig.version),
		...getLastOptions(config.uploadOptions),
	};

	const previewOptions = {
		...commonConfig,
		...getLastOptions(config.previewOptions),
	};

	return {
		projectOptions,
		uploadOptions,
		previewOptions,
		plugins: config.plugins,
		replaceRules: config.replaceRules,
		name: config.name,
		// id: uniqueId(),
	};
};

// 获取配置的列表
export const getConfigList = async (configPath: string) => {
	let configPathList: string[] = [];
	try {
		configPathList = fs.readdirSync(configPath);
	} catch (error) {
		logger.error(error);
		throw new Error('请设置小程序配置路径 configPath');
	}
	if (configPathList.length === 0) {
		throw new Error('请设置小程序配置路径 configPath');
	}

	// logger.log(configPathList, 'configPathList');
	const configList: any[] = [];
	for (let el of configPathList) {
		let config = await resolveFileConfig(`${configPath}/${el}`);
		configList.push(config);
	}

	return configList;
};

// 解析配置
export const resolveConfig = async (config: InlineConfig) => {
	// resolve root
	const resolvedRoot = getResolvedRoot(config);

	let loadResult;
	// 如果指定了配置的文件
	if (config.config) {
		loadResult = await resolveFileConfig(
			path.resolve(resolvedRoot, config.config),
		);
	} else {
		loadResult = await loadConfigFromFile(resolvedRoot);
	}

	if (!isObject(loadResult)) {
		const envFile = config.mode ? `.env.${config.mode}` : '.env';
		logger.log(
			`加载 mini.config.(js|json) 失败，或未指定可用的配置文件，将使用 ${envFile} 中的配置`,
			loadResult,
		);
		// @ts-ignore
		loadResult = loadEnv(config.mode, resolvedRoot);
	}
	if (!isObject(loadResult)) {
		throw new Error(
			'未发现配置项，请检查 mini-ci 的 mini.config.js 或者 .env 相关配置',
		);
	}

	let loadResultList: any[] = [];
	// 单选或者多选
	if (config.useSelect || config.useMultiSelect || config.useAllConfig) {
		const configList = await getConfigList(
			path.resolve(resolvedRoot, loadResult?.configPath!),
		);
		const allConfigList = [...configList, loadResult];
		// 批量操作所有
		if (config.useAllConfig) {
			loadResultList = allConfigList;
		} else {
			loadResultList = await select({
				configList: allConfigList,
				useSelect: config.useSelect,
				useMultiSelect: config.useMultiSelect,
			});
		}
	} else {
		loadResultList = [loadResult];
	}

	const lastConfigList = loadResultList.map((item) => {
		return mergeConfig(config, item!);
	});

	return lastConfigList;
};
