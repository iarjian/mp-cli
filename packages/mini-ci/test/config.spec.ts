import { expect, test } from 'vitest';
import { resolveConfig } from '../src/config';
// import path from 'node:path';
// @ts-ignore
import miniConfig from '../mini.config.js';

test('resolveConfig miniConfig', async () => {
	const configList = await resolveConfig({});

	const config = configList[0] || {};

	expect(config?.projectOptions?.appid).toBe(miniConfig.appid);
	expect(config.plugins).toStrictEqual(miniConfig.plugins);
	expect(config.replaceRules).toStrictEqual(miniConfig.replaceRules);
});

test('resolveConfig toMatchSnapshot', async () => {
	const configList = await resolveConfig({});
	expect(configList).toMatchSnapshot();
});

test('resolveConfig -c --config miniConfigList/example.js', async () => {
	const configList = await resolveConfig({
		config: 'miniConfigList/example.js',
	});
	expect(configList).toMatchSnapshot();
});

test('resolveConfig -c --config miniConfigList/example.json', async () => {
	const configList = await resolveConfig({
		config: 'miniConfigList/example.json',
	});
	expect(configList).toMatchSnapshot();
});
