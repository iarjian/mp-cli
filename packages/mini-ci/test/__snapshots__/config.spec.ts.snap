// Vitest Snapshot v1

exports[`resolveConfig --config miniConfigList/example.js 1`] = `
[
  {
    "name": "若川视野-小程序-example.js",
    "plugins": [],
    "previewOptions": {
      "robot": undefined,
      "setting": {
        "es6": true,
        "es7": true,
        "ignoreUploadUnusedFiles": true,
        "minify": true,
      },
      "version": "1.1.0",
    },
    "projectOptions": {
      "appid": "wxdd3948dc1c7f95c2",
      "ignores": [
        "node_modules/**/*",
        "CHANGELOG.md",
        "README.md",
        "yarn.lock",
        "package-lock.json",
        "pnpm-lock.yaml",
      ],
      "privateKeyPath": "./key/private.wxdd3948dc1c7f95c2.key",
      "projectPath": "../../../tdesign-miniprogram-starter-retail",
      "type": "miniProgram",
    },
    "replaceRules": [],
    "uploadOptions": {
      "desc": "v1.1.0 - ccb6cbe - by@若川",
      "robot": undefined,
      "setting": {
        "es6": true,
        "es7": true,
        "ignoreUploadUnusedFiles": true,
        "minify": true,
      },
      "version": "1.1.0",
    },
  },
]
`;

exports[`resolveConfig --config miniConfigList/example.json 1`] = `
[
  {
    "name": "若川视野的小程序-默认json",
    "plugins": undefined,
    "previewOptions": {
      "robot": undefined,
      "setting": {
        "es6": true,
        "es7": true,
        "ignoreUploadUnusedFiles": true,
        "minify": true,
      },
      "version": "1.1.0",
    },
    "projectOptions": {
      "appid": "wxdd3948dc1c7f95c2",
      "ignores": [
        "node_modules/**/*",
        "CHANGELOG.md",
        "README.md",
        "yarn.lock",
        "package-lock.json",
        "pnpm-lock.yaml",
      ],
      "privateKeyPath": "./key/private.wxdd3948dc1c7f95c2.key",
      "projectPath": "../../../tdesign-miniprogram-starter-retail",
      "type": "miniProgram",
    },
    "replaceRules": undefined,
    "uploadOptions": {
      "desc": "v1.1.0 - ccb6cbe - by@若川",
      "robot": undefined,
      "setting": {
        "es6": true,
        "es7": true,
        "ignoreUploadUnusedFiles": true,
        "minify": true,
      },
      "version": "1.1.0",
    },
  },
]
`;

exports[`resolveConfig -c --config miniConfigList/example.js 1`] = `
[
  {
    "name": "若川视野-小程序-example.js",
    "plugins": [],
    "previewOptions": {
      "robot": undefined,
      "setting": {
        "es6": true,
        "es7": true,
        "ignoreUploadUnusedFiles": true,
        "minify": true,
      },
      "version": "1.1.0",
    },
    "projectOptions": {
      "appid": "wxdd3948dc1c7f95c2",
      "ignores": [
        "node_modules/**/*",
        "CHANGELOG.md",
        "README.md",
        "yarn.lock",
        "package-lock.json",
        "pnpm-lock.yaml",
      ],
      "privateKeyPath": "./key/private.wxdd3948dc1c7f95c2.key",
      "projectPath": "../../../tdesign-miniprogram-starter-retail",
      "type": "miniProgram",
    },
    "replaceRules": [],
    "uploadOptions": {
      "desc": "v1.1.0 - ccb6cbe - by@若川",
      "robot": undefined,
      "setting": {
        "es6": true,
        "es7": true,
        "ignoreUploadUnusedFiles": true,
        "minify": true,
      },
      "version": "1.1.0",
    },
  },
]
`;

exports[`resolveConfig -c --config miniConfigList/example.json 1`] = `
[
  {
    "name": "若川视野的小程序-默认json",
    "plugins": undefined,
    "previewOptions": {
      "robot": undefined,
      "setting": {
        "es6": true,
        "es7": true,
        "ignoreUploadUnusedFiles": true,
        "minify": true,
      },
      "version": "1.1.0",
    },
    "projectOptions": {
      "appid": "wxdd3948dc1c7f95c2",
      "ignores": [
        "node_modules/**/*",
        "CHANGELOG.md",
        "README.md",
        "yarn.lock",
        "package-lock.json",
        "pnpm-lock.yaml",
      ],
      "privateKeyPath": "./key/private.wxdd3948dc1c7f95c2.key",
      "projectPath": "../../../tdesign-miniprogram-starter-retail",
      "type": "miniProgram",
    },
    "replaceRules": undefined,
    "uploadOptions": {
      "desc": "v1.1.0 - ccb6cbe - by@若川",
      "robot": undefined,
      "setting": {
        "es6": true,
        "es7": true,
        "ignoreUploadUnusedFiles": true,
        "minify": true,
      },
      "version": "1.1.0",
    },
  },
]
`;

exports[`resolveConfig toMatchSnapshot 1`] = `
[
  {
    "name": "若川视野-小程序1-mini.config.js",
    "plugins": [],
    "previewOptions": {
      "robot": undefined,
      "setting": {
        "es6": true,
        "es7": true,
        "ignoreUploadUnusedFiles": true,
        "minify": true,
      },
      "version": "1.1.0",
    },
    "projectOptions": {
      "appid": "wxdd3948dc1c7f95c2",
      "ignores": [
        "node_modules/**/*",
        "CHANGELOG.md",
        "README.md",
        "yarn.lock",
        "package-lock.json",
        "pnpm-lock.yaml",
      ],
      "privateKeyPath": "./key/private.wxdd3948dc1c7f95c2.key",
      "projectPath": "../../../tdesign-miniprogram-starter-retail",
      "type": "miniProgram",
    },
    "replaceRules": [],
    "uploadOptions": {
      "desc": "v1.1.0 - ccb6cbe - by@若川",
      "robot": undefined,
      "setting": {
        "es6": true,
        "es7": true,
        "ignoreUploadUnusedFiles": true,
        "minify": true,
      },
      "version": "1.1.0",
    },
  },
]
`;
